contracts/smartwallet/BoltzSmartWallet.sol

```diff
@@ -0,0 +1,330 @@
+        }
+    }
+
+    function _verifySig(
+        bytes32 suffixData,
+        ForwardRequest memory req,
+        bytes memory sig
+    ) internal view {
+        //Verify Owner
+        require(
+            getOwner() == keccak256(abi.encodePacked(req.from)),
+            "Not the owner of the SmartWallet"
+        );
+
+        //Verify nonce
+        require(nonce == req.nonce, "nonce mismatch");
+
+        require(
+            RSKAddrValidator.safeEquals(
+                keccak256(
+                    abi.encodePacked(
+                        "\x19\x01",
+                        domainSeparator,
+                        keccak256(_getEncoded(suffixData, req))
+                    )
+                ).recover(sig),
+                req.from
+            ),
+            "Signature mismatch"
+        );
+    }
+
+    function _getEncoded(
+        bytes32 suffixData,
+        ForwardRequest memory req
+    ) internal pure returns (bytes memory) {
+        return
+            abi.encodePacked(
+                keccak256(
+                    "RelayRequest(address relayHub,address from,address to,address tokenContract,uint256 value,uint256 gas,uint256 nonce,uint256 tokenAmount,uint256 tokenGas,uint256 validUntilTime,bytes data,RelayData relayData)RelayData(uint256 gasPrice,address feesReceiver,address callForwarder,address callVerifier)"
+                ), //requestTypeHash,
+                abi.encode(
+                    req.relayHub,
+                    req.from,
+                    req.to,
+                    req.tokenContract,
+                    req.value,
+                    req.gas,
+                    req.nonce,
+                    req.tokenAmount,
+                    req.tokenGas,
+                    req.validUntilTime,
+                    keccak256(req.data)
+                ),
+                suffixData
+            );
+    }
+
+    function isInitialized() external view returns (bool) {
+        if (getOwner() == bytes32(0)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * This Proxy will first charge for the deployment and then it will pass the
+     * initialization scope to the wallet logic.
+     * This function can only be called once, and it is called by the Factory during deployment
+     * @param owner - The EOA that will own the smart wallet
+     * @param tokenContract - Token used for payment of the deploy
+     * @param feesReceiver - Recipient of payment
+     * @param tokenAmount - Amount to pay
+     * @param tokenGas - Gas limit of payment
+     * @param to - Destination contract to execute
+     * @param value - Value to send to destination contract
+     * @param destinationCallGas - Gas limit of destination contract
+     * @param data - Data to be execute by destination contract
+     */
+    function initialize(
+        address owner,
+        address tokenContract,
+        address feesReceiver,
+        uint256 tokenAmount,
```

Comment:
The initialize function is not part of an Interface. The name and number of attributes could be reshaped to the Boltz Scenario. In Boltz, a user does not pay with Tokens. If we want to have an option that has all the possible scenarios (pay with tokens, or RBTC), that's fine but then we need another template that is minimalistic and only implements what Boltz is requiring (the deploy of the smart wallet is paid with the RBTC from the claim function). Then they can choose whatever template fits them best (they could, for example, choose the minimalistic one if it is much cheaper to execute, as it requires less parameters and less opcodes)

Reply:
updated